arangod/RocksDBEngine/RocksDBReplicationContext.cpp

```diff
@@ -69,170 +68,138 @@ TRI_voc_cid_t normalizeIdentifier(transaction::Methods const& trx,
-      _collection->release();
-    }
-
-    _collection = getCollectionIterator(id, /*sorted*/ true);
-    if (nullptr == _collection) {
-      return TRI_ERROR_BAD_PARAMETER;
+/// remove matching iterator
+void RocksDBReplicationContext::releaseIterators(TRI_vocbase_t& vocbase, TRI_voc_cid_t cid) {
+  MUTEX_LOCKER(locker, _contextLock);
+  auto it = _iterators.find(cid);
+  if (it != _iterators.end()) {
+    if (it->second->isUsed()) {
+      LOG_TOPIC(ERR, Logger::REPLICATION) << "trying to delete used iterator";
+    } else {
+      _iterators.erase(it);
     }
+  } else {
+    LOG_TOPIC(ERR, Logger::REPLICATION) << "trying to delete non-existent iterator";
   }
-
-  return TRI_ERROR_NO_ERROR;
 }
 
-int RocksDBReplicationContext::chooseDatabase(TRI_vocbase_t& vocbase) {
-  TRI_ASSERT(_users > 0);
-  MUTEX_LOCKER(locker, _contextLock);
-
-  if (&(_guard->database()) == &vocbase) {
-    return TRI_ERROR_NO_ERROR;  // nothing to do here
+/// Bind collection for incremental sync
+std::tuple<Result, TRI_voc_cid_t, uint64_t>
+RocksDBReplicationContext::bindCollectionIncremental(TRI_vocbase_t& vocbase,
+                                                     std::string const& cname) {
+  
+  std::shared_ptr<LogicalCollection> logical{
+    vocbase.lookupCollection(cname)
+  };
+  
+  if (!logical) {
+    return std::make_tuple(TRI_ERROR_BAD_PARAMETER, 0, 0);
   }
+  TRI_voc_cid_t cid = logical->id();
 
-  // need to actually change it, first make sure we're alone in this context
-  if (_users > 1) {
-    return TRI_ERROR_CURSOR_BUSY;
+  MUTEX_LOCKER(writeLocker, _contextLock);
+  
+  auto it = _iterators.find(cid);
+  if (it != _iterators.end()) { // nothing to do here
+    return std::make_tuple(Result{}, it->second->logical->id(),
+                           it->second->numberDocuments);
   }
-
-  // make the actual change
-  internalBind(vocbase, true);
-
-  return TRI_ERROR_NO_ERROR;
+  
+  uint64_t numberDocuments;
+  bool isNumberDocsExclusive = false;
+  auto* rcoll = static_cast<RocksDBCollection*>(logical->getPhysical());
+  if (_snapshot == nullptr) {
+    // fetch number docs and snapshot under exclusive lock
+    // this should enable us to correct the count later
+    auto lockGuard = scopeGuard([rcoll] { rcoll->unlockWrite(); });
+    int res = rcoll->lockWrite(transaction::Options::defaultLockTimeout);
+    if (res != TRI_ERROR_NO_ERROR) {
+      lockGuard.cancel();
+      return std::make_tuple(res, 0, 0);
+    }
+    
+    numberDocuments = rcoll->numberDocuments();
+    isNumberDocsExclusive = true;
+    lazyCreateSnapshot();
+    lockGuard.fire(); // release exclusive lock
+  } else { // fetch non-exclusive
+    numberDocuments = rcoll->numberDocuments();
+    lazyCreateSnapshot();
+  }
+
+  auto iter = std::make_unique<CollectionIterator>(vocbase, logical, true, _snapshot);
+  auto result = _iterators.emplace(cid, std::move(iter));
+  if (!result.second) {
+    return std::make_tuple(TRI_ERROR_OUT_OF_MEMORY, 0, 0);
```

Comment:
If the bool is false, it means it wasn't inserted because an entry with the same key already exists in the map. In case of OOM, you need to catch an `std::bad_alloc`.

Reply:
@graetzer This is still dangling...